gson/src/test/java/com/google/gson/JsonPrimitiveTest.java

```diff
@@ -169,6 +169,25 @@ public void testLongEqualsBigInteger() {
     assertEquals(p1.hashCode(), p2.hashCode());
   }
 
+  public void testHashCodeBigIntegerEqualsBigInteger() {
+    // within long range
+    BigInteger smallBigInt = new BigInteger("10");
+    JsonPrimitive p1 = new JsonPrimitive(smallBigInt);
+    JsonPrimitive p2 = new JsonPrimitive(smallBigInt);
+    assertTrue("should be different objects: " + p1 + ", " + p2, p1 != p2);
+    assertTrue("p1 and p2 should have the same value", p1.equals(p2));
+    assertTrue("p1 and p2 should have the same hash code", p1.hashCode() == p2.hashCode());
+
+    // out of long range
+    BigInteger limitPlus5 =  // 2^64 + 5
+      new BigInteger("18446744073709551616").add(BigInteger.valueOf(5));
+    JsonPrimitive p3 = new JsonPrimitive(limitPlus5);
+    JsonPrimitive p4 = new JsonPrimitive(limitPlus5);
+    assertTrue("should be different objects: " + p3 + ", " + p4, p3 != p4);
+    assertTrue("p3 and p4 should have the same value", p3.equals(p4));
+    assertTrue("p3 and p4 should have the same hash code", p3.hashCode() == p4.hashCode());
```

Comment:
Use `assertEquals(int, int)` to get a more descriptive error message in case of mismatch